File: DragonIsland-PSP/SoundEffect.cpp
#include "SoundEffect.h"

SoundEffect::SoundEffect()
{
	mSound[0] = oslLoadSoundFileWAV("res/sfx/block.wav", OSL_FMT_NONE);
	mSound[1] = oslLoadSoundFileWAV("res/sfx/coin.wav", OSL_FMT_NONE);
	mSound[2] = oslLoadSoundFileWAV("res/sfx/die.wav", OSL_FMT_NONE);
	mSound[3] = oslLoadSoundFileWAV("res/sfx/fire.wav", OSL_FMT_NONE);
	mSound[4] = oslLoadSoundFileWAV("res/sfx/hit.wav", OSL_FMT_NONE);
	mSound[5] = oslLoadSoundFileWAV("res/sfx/jump.wav", OSL_FMT_NONE);
	mSound[6] = oslLoadSoundFileWAV("res/sfx/kill.wav", OSL_FMT_NONE);
	mSound[7] = oslLoadSoundFileWAV("res/sfx/option.wav", OSL_FMT_NONE);
	mSound[8] = oslLoadSoundFileWAV("res/sfx/powerdown.wav", OSL_FMT_NONE);
	mSound[9] = oslLoadSoundFileWAV("res/sfx/powerup.wav", OSL_FMT_NONE);
	mSound[10] = oslLoadSoundFileWAV("res/sfx/select.wav", OSL_FMT_NONE);
	mSound[11] = oslLoadSoundFileWAV("res/sfx/speech.wav", OSL_FMT_NONE);
	mSound[12] = oslLoadSoundFileWAV("res/sfx/tank.wav", OSL_FMT_NONE);
}

void SoundEffect::play(string filename, int repeat)
{
	if (Settings::Sound) {
		if (filename == "block.wav")
			oslPlaySound(mSound[0], 0);
		else if (filename == "coin.wav")
			oslPlaySound(mSound[1], 0);
		else if (filename == "die.wav")
			oslPlaySound(mSound[2], 0);
		else if (filename == "fire.wav")
			oslPlaySound(mSound[3], 0);
		else if (filename == "hit.wav")
			oslPlaySound(mSound[4], 0);
		else if (filename == "jump.wav")
			oslPlaySound(mSound[5], 1);
		else if (filename == "kill.wav")
			oslPlaySound(mSound[6], 0);
		else if (filename == "option.wav")
			oslPlaySound(mSound[7], 0);
		else if (filename == "powerdown.wav")
			oslPlaySound(mSound[8], 0);
		else if (filename == "powerup.wav")
			oslPlaySound(mSound[9], 0);
		else if (filename == "select.wav")
			oslPlaySound(mSound[10], 0);
		else if (filename == "speech.wav")
			oslPlaySound(mSound[11], 0);
		else if (filename == "tank.wav")
			oslPlaySound(mSound[12], 0);
	}
}

File: DragonIsland-Android-Eclipse/src/com/citex/dragonisland/android/audio/SoundEffect.java
package com.citex.dragonisland.android.audio;

import java.io.IOException;
import java.util.HashMap;

import android.annotation.SuppressLint;
import android.content.res.AssetFileDescriptor;
import android.media.AudioManager;
import android.media.SoundPool;

import com.citex.dragonisland.android.MainActivity;
import com.citex.dragonisland.core.audio.ISoundEffect;
import com.citex.dragonisland.core.game.Settings;

/**
 * This class plays wav sounds effects.
 * @author Lawrence Schmid
 */
public class SoundEffect implements ISoundEffect {

	/** Sound pool. */
	private static SoundPool mSoundPool;
	
	/** Sound pool map. */
	private static HashMap<Integer, Integer> mSoundPoolMap;
	
	/**
	 * Initialise a collection of sound effects.
	 * @param path Path to the wav sound effect files.
	 * @throws IOException 
	 * @throws IllegalStateException 
	 */
	@SuppressLint("UseSparseArrays") 
	public SoundEffect(String path) throws IllegalStateException, IOException {

		// Initialise the sound pool.
		mSoundPool = new SoundPool(4, AudioManager.STREAM_MUSIC, 100);
		
		// Sound pool map.
		mSoundPoolMap = new HashMap<Integer, Integer>(12);  
		
		// Load sound effects from assets.
		AssetFileDescriptor afd = MainActivity.mAssetManager.openFd(path + "block.wav");
		mSoundPoolMap.put(1, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "coin.wav");
		mSoundPoolMap.put(2, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "die.wav");
		mSoundPoolMap.put(3, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "fire.wav");
		mSoundPoolMap.put(4, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "hit.wav");
		mSoundPoolMap.put(5, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "jump.wav");
		mSoundPoolMap.put(6, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "kill.wav");
		mSoundPoolMap.put(7, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "option.wav");
		mSoundPoolMap.put(8, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "powerdown.wav");
		mSoundPoolMap.put(9, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "powerup.wav");
		mSoundPoolMap.put(10, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "select.wav");
		mSoundPoolMap.put(11, mSoundPool.load(afd, 1));

	}

	/**
	 * Plays a wav sound effect 
	 * @param soundName File name of the sound effect.
	 */
	public void play(String soundName) {
		
		// Initialise sound properties.
		float leftVolume = 0.5f;
		float rightVolume = 0.5f;
		int priority = 1;
		int loop = 0;
		float rate = 1.0f;
		
		if (Settings.Sound) {

			if (soundName.equals("block.wav"))
				mSoundPool.play(mSoundPoolMap.get(1), leftVolume, rightVolume, priority, loop, rate);
			else if (soundName.equals("coin.wav"))
				mSoundPool.play(mSoundPoolMap.get(2), leftVolume - 0.05f, rightVolume - 0.05f, priority, loop, rate);
			else if (soundName.equals("die.wav"))
				mSoundPool.play(mSoundPoolMap.get(3), leftVolume - 0.05f, rightVolume - 0.05f, priority, loop, rate);
			else if (soundName.equals("fire.wav"))
				mSoundPool.play(mSoundPoolMap.get(4), leftVolume, rightVolume, priority, loop, rate);
			else if (soundName.equals("hit.wav"))
				mSoundPool.play(mSoundPoolMap.get(5), leftVolume, rightVolume, priority, loop, rate);
			else if (soundName.equals("jump.wav"))
				mSoundPool.play(mSoundPoolMap.get(6), leftVolume - 0.1f, rightVolume - 0.1f, priority, loop, rate);
			else if (soundName.equals("kill.wav"))
				mSoundPool.play(mSoundPoolMap.get(7), leftVolume, rightVolume, priority, loop, rate);
			else if (soundName.equals("option.wav"))
				mSoundPool.play(mSoundPoolMap.get(8), leftVolume, rightVolume, priority, loop, rate);
			else if (soundName.equals("powerdown.wav"))
				mSoundPool.play(mSoundPoolMap.get(9), leftVolume, rightVolume, priority, loop, rate);
			else if (soundName.equals("powerup.wav"))
				mSoundPool.play(mSoundPoolMap.get(10), leftVolume, rightVolume, priority, loop, rate);
			else if (soundName.equals("select.wav"))
				mSoundPool.play(mSoundPoolMap.get(11), leftVolume, rightVolume, priority, loop, rate);
			
		}
	}

    /**
     * Destroys the resources.
     */
	public void destroy() {
		
		// Destroy sound effects.
		mSoundPool.release();
		
	}
}
